libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/builder/NavRouteOptionsBuilder.kt

```diff
@@ -0,0 +1,362 @@
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface NoWaypointsOptionsBuilder {
+    fun fromStartLocation(
+        point: Point,
+        bearing: Double? = null,
+        zLevel: Int?
+    ): WaypointsInProgressBuilder
+
+    fun fromCurrentLocation(): WaypointsInProgressBuilder
+}
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface WaypointsInProgressBuilder {
+    fun addIntermediateWaypoint(
+        coordinate: Point,
+        name: String? = null,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+        targetCoordinate: Point? = null
+    ): WaypointsInProgressBuilder
+
+    fun addIntermediateSilentWaypoint(
+        coordinate: Point,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+    ): WaypointsInProgressBuilder
+
+    fun toDestination(
+        coordinate: Point,
+        name: String? = null,
+        bearing: Double? = null,
+        zLevel: Int? = null,
+        targetCoordinate: Point? = null
+    ): RouteOptionsBuilderWithWaypoints
+}
+
+@ExperimentalPreviewMapboxNavigationAPI
+interface RouteOptionsBuilderWithWaypoints {
+    fun profileDriving(
+        drivingSpecificSetup: DrivingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileDrivingTraffic(
+        drivingSpecificSetup: DrivingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileWalking(
+        walkingSpecificSetup: WalkingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+
+    fun profileCycling(
+        cyclingSpecificSetup: CyclingSpecificSetup.() -> Unit = { }
+    ): RouteOptionsBuilderWithWaypoints
+}
+
+@ExperimentalPreviewMapboxNavigationAPI
+internal class NavRouteOptionsBuilder internal constructor(
+    private val locationProvider: LocationProvider
+) : NoWaypointsOptionsBuilder, WaypointsInProgressBuilder, RouteOptionsBuilderWithWaypoints {
+
+    private lateinit var destination: Waypoint
+    private val builder = RouteOptions.builder()
+    private var locationFrom: CurrentLocation? = null
+    private val intermediateWaypoints = mutableListOf<Waypoint>()
+    private var profile: String? = null
+
+    override fun fromCurrentLocation(): WaypointsInProgressBuilder {
+        return this
+    }
+
+    override fun toDestination(
+        coordinate: Point,
+        name: String?,
+        bearing: Double?,
+        zLevel: Int?,
+        targetCoordinate: Point?,
+    ): RouteOptionsBuilderWithWaypoints {
+        destination = Waypoint(
+            coordinate = coordinate,
+            name = name,
+            bearing = bearing,
+            zLevel = zLevel,
+            targetCoordinate = targetCoordinate,
+        )
+        return this
+    }
+
+    internal fun applyLanguageAndVoiceUnitOptions(context: Context): NavRouteOptionsBuilder {
+        builder.applyLanguageAndVoiceUnitOptions(context)
+        return this
+    }
+
+    internal suspend fun build(): RouteOptions {
```

Comment:
Builder a **suspend** method? I understand it's because of `locationProvider.getCurrentLocation()` but I don't think builders should ever be **suspend** methods.

To eliminate the need to call `getCurrentLocation()`, we could set a placeholder object that will tell MapboxNavigation to use current location when `requestRoutes()` is called.

eg. placeholder object could look like this
```kotlin
data class Waypoint(
    val coordinate: Point,
    val name: String?,
    val bearing: Double?,
    val zLevel: Int?,
    val isSilent: Boolean = false,
    val targetCoordinate: Point?
) {
    companion object {
        val CURRENT_LOCATION = Waypoint(
            coordinate = Point.fromLngLat(Double.NaN, Double.NaN),
            name = "CURRENT_LOCATION",
            bearing = null,
            zLevel = null,
            isSilent = false,
            targetCoordinate = null
        )
    }
}
```

The current location retrieval could be customized by additionally passing `LocationProvider` to RouteOptions and updating either `MapboxNavigation.requestRoutes()` or `MapboxDirectionsSession.requestRoutes()` to launch coroutine and acquiring waypoint location before forwarding it to the router.

Reply:
>  I don't think builders should ever be suspend methods.

Why do you think so? Is there any unexpected consequences I can get violating "build should be synchronous" rule?

Reply:
Just a note. I don't like current API in terms of asynchronously. Currently the API makes a user to asynchronously build a `RouteOptions` object first and then request again asynchronously a route. It would be a callback hell without coroutines 🙂 
I don't know good solution for this right now, so I decided to leave it for the next iterations.

Reply:
I'd like to propose a simple solution for this problem - expose a new `MapboxNavigation#requestRoutes(NavRouteOptionsBuilder)` which will build the options and request a route in one asynchronous call.

We should still leave the `buildRouteOptions` public for developers that would like to mutate the predefined options but the basic flow would be not to build the options manually.

Reply:
@LukasPaczos , I agree that`MapboxNavigation#requestRoutes(NavRouteOptionsBuilder)` is the best option. The only reason I didn't do that is that because it requires a tricky cancellation implementation. I wanted the first iteration to be shorter and postponed cancellation to the next iterations. 

But actually, cancellation won't be complex if `MapboxNavigation#requestRoutes(NavRouteOptionsBuilder)` will be a suspend function. Will try it.

Reply:
Implemented `MapboxNavigation#requestRoutes(NavRouteOptionsBuilder)` as suspend function